File: app/src/main/java/com/example/mvvm_artichecture_sample/base/network/RetrofitClient.java
package com.example.mvvm_artichecture_sample.base.network;

import com.example.mvvm_artichecture_sample.base.network.base.BaseRetrofitClient;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;

public class RetrofitClient extends BaseRetrofitClient {
    private static final long WRITE_TIME_OUT = 10;
    private static final long READ_TIME_OUT = 20;
    public static final String BASE_URL = "https://api.printful.com/";
    private static RetrofitClient instance = null;


    private RetrofitClient() {

    }

    public static RetrofitClient getInstance() {
        if (instance == null) {
            instance = new RetrofitClient();
        }
        return instance;
    }

    @Override
    protected OkHttpClient getClient() {
        return getOkHttpClient();
    }

    private OkHttpClient getOkHttpClient() {


        OkHttpClient.Builder builder = new OkHttpClient.Builder()
                .readTimeout(READ_TIME_OUT, TimeUnit.SECONDS)
                .retryOnConnectionFailure(false);
        return builder.build();
    }

    @Override
    protected String getBaseUrl() {
        return BASE_URL;
    }
}

File: app/src/main/java/com/example/mvvm_artichecture_sample/base/network/RetrofitClient.kt
package com.example.mvvm_artichecture_sample.base.network

import com.example.mvvm_artichecture_sample.base.network.base.BaseRetrofitClient

import java.util.concurrent.TimeUnit

import okhttp3.OkHttpClient

class RetrofitClient private constructor() : BaseRetrofitClient() {

    override val client: OkHttpClient
        get() = okHttpClient

    private val okHttpClient: OkHttpClient
        get() {


            val builder = OkHttpClient.Builder()
                    .readTimeout(READ_TIME_OUT, TimeUnit.SECONDS)
                    .retryOnConnectionFailure(false)
            return builder.build()
        }

    override val baseUrl: String
        get() = BASE_URL

    companion object {
        private val WRITE_TIME_OUT: Long = 10
        private val READ_TIME_OUT: Long = 20
        val BASE_URL = "https://api.printful.com/"
        private var instance: RetrofitClient? = null

        fun getInstance(): RetrofitClient {
            if (instance == null) {
                instance = RetrofitClient()
            }
            return instance!!
        }
    }
}
